### ISSUE TITLE
Is Sersic2D oversampled at its center?

### ISSUE THREAD

**Issue by canorve** [astropy/astropy#11179]
Sersic function changes quickly at the center for large Sersic indexes. This makes the function very steep at the center. To correct this,  pixels near the center of the model must be oversampled. 

This is explained in section 2.3.1 of [https://ui.adsabs.harvard.edu/link_gateway/2002AJ....124..266P/doi:10.1086/340952](url)  

and section 2.1 of  [https://ui.adsabs.harvard.edu/link_gateway/2007ApJS..172..615H/doi:10.1086/518836](url)

Looking at the code, it seems it might not.

**Comment by pllim** [comment]
> it seems it might not

Hello and thanks for reporting this. Do you have a use case that ran into this problem? If so, do you have a reproducible code snippet to share?

**Comment by canorve** [comment]
Hello pllim, No I don't.  But I think @BorisHaeussler might give us some insight

**Comment by BorisHaeussler** [comment]
Wow, that email came out of the blue ;-)
(I don't use the python routine in question, so I am making general comments)

I have plenty of examples, and this is a well known issue with most routines (although not documented very well), the problem can be on a level of ~10% flux in the central pixels, especially for small, high-n objects.

The problem is the curvature of the profile. Most routines work well for the center of the pixels. But the flux in a pixel is not equal to the central flux times pixel area, because the profile is curved.

I have written a little about it in my paper in 2007, https://arxiv.org/pdf/0704.2601.pdf (I think that's the link above, but that doesn't work for me)
In Section 2.1, I also explain my (cheat) workaround (although I changed this meanwhile, I now oversample in 3 steps - 4 steps in total - the central part by a factor of 1000). 
If you wanted to do this properly, you'd have to numerically integrate the profile over the pixel area, but that's very CPU intensive. Instead, I create objects (much) larger and shrink it down, which gives satisfactory results while being MUCH faster.

I'd be happy to provide my own code if you want to have a look, but it's written in IDL, so not very useful directly.
Basically, it's creating 4 profiles, each oversampled by a larger factor, bins them down and makes sure that they are aligned well to avoid edge issues.

**Comment by dhomeier** [comment]
Thanks for the clarification @BorisHaeussler! In the paper however you wrote
> a factor of 100 gain in speed compared to the analytical integration

– indeed would not numerically integrating essentially just amount to evaluating the profile oversampled at appropriate steps? These seem to be the two cases compared in https://github.com/astropy/astropy/issues/11178#issue-771297970 where `integrate.dblquad` automatically selects the required sampling for the desired precision, right?

Whether this is a bug in `Sersic2D` depends on whether the model is assumed to sample the profile at the given `x[:]`, `y[:]`, or represent a binned image – this is not very clear for a number of models, but the latter would certainly be the more typical case for analysing real image data.

**Comment by BorisHaeussler** [comment]
> – indeed would not numerically integrating essentially just amount to evaluating the profile oversampled at appropriate steps? These seem to be the two cases compared in [#11178 (comment)] where `integrate.dblquad` automatically selects the required sampling for the desired precision, right?

I that comment is referring to the python routine which I don't use.
From what I understand, that comment talks about the total magnitude of the profile, not a specific pixel. So it's something else. (While this is unlikely), the total flux might well be correct, but the distribution might still be wrong.

I am talking about integrating the profile shape not over all radii (as in #11178), but over a pixel area. This CAN be done, but is very slow. I have done the test of comparing the analytic profile to the oversampled (and not) profiles I create using my IDL code in the past (~2008?) (and a non-oversampled python code will likely have the same issue), and my comment in my paper was referring to that test.
Indeed, what I did was create profiles at different sizes, sersic indices, etc, integrated pixel by pixel. And then subtract/divide them from the ones I got from my routine. They agreed within 0.03% at the end, I think to remember. Without the oversampling, the difference was up to 10% for individual pixels.
 
> Whether this is a bug in `Sersic2D` depends on whether the model is assumed to sample the profile at the given `x[:]`, `y[:]`, or represent a binned image – this is not very clear for a number of models, but the latter would certainly be the more typical case for analysing real image data.

No, I disagree. In fact, both, the value at x/y and a binned image will be wrong.
Yes, it might not be a 'bug', because the code does what it's been told to do. It's just been told to do the wrong thing.

The profiles that you create without oversampling (or better: integrating) do NOT describe a sersic function in the central pixels (which are absolutely critical to e.g. light profile fitting, as they contain the most light), and also not in what you call a 'binned' image.
A binned image in in fact the integrated profile over a pixel area, because that's what a detector actually does. Your routine should try to get as close to that as possible (in a reasonable time).

A wrong profile shape makes them useless for many purposes (e.g. simulated artificial images as the OP seems to try and do).
In my papers, I can continuously explain peoples finding by their usage of such bad routines.

Sure, the total magnitude of such an object might be ok, and the result might be useable for specific purposes, but then you're going the way or IRAF/MKOBJECT, which has been badly used in the literature and created some bad results. I discuss it somehere.
In short, IRAF/MKOBJECT gets the flux out to a certain radius, and then fold the missing flux (outside that radius) into the central few pixels, destroying the light profile. This is great to test aperture photometry, but it's a disaster when you talk about profile shapes.
Without caring about the details, a user would use this routine and rely on it being good.

astropy should not go that way.

**Comment by dhomeier** [comment]
> I am talking about integrating the profile shape not over all radii (as in #11178), but over a pixel area. This CAN be done, but is very slow. I have done the test of comparing the analytic profile to the oversampled (and not) profiles I create using my IDL code in the past (~2008?) (and a non-oversampled python code will likely have the same issue), and my comment in my paper was referring to that test.

But you were comparing to analytic integration, not numerical? That was the point I wanted to have clarity on. In principle the Scipy integrate function `dblquad` could be used on every pixel, but I have no idea if the performance would be acceptable. If this already had been shown to be 100 x slower, it would clearly not be an option, but otherwise probably needs further testing to see if it could be an alternative to oversampling.

> 
> > Whether this is a bug in `Sersic2D` depends on whether the model is assumed to sample the profile at the given `x[:]`, `y[:]`, or represent a binned image – this is not very clear for a number of models, but the latter would certainly be the more typical case for analysing real image data.
> 
> No, I disagree. In fact, both, the value at x/y and a binned image will be wrong.

So you are saying the correct intensity at the exact location x, y (radius `r`, if this could indeed be measured somehow) is _not_ given by 
`Ie * np.exp(-bn * (r ** (1 / n) - 1))` ?

> Yes, it might not be a 'bug', because the code does what it's been told to do. It's just been told to do the wrong thing.

Well, it should return the intensity distribution as defined in the reference http://ned.ipac.caltech.edu/level5/March05/Graham/Graham2.html – if it doesn't even do that for sampled intensities, that is clearly a bug. That it is not producing binned fluxes should at least be clarified in the documentation, but ideally I think models should provide either option, as it is e.g. implemented now for `Spectrum1D`. But that would require some major work on the entire `modeling` package.
But from your comments above I take that this particular model should be taken care of by a galaxy person...

**Comment by BorisHaeussler** [comment]
> But you were comparing to analytic integration, not numerical? That was the point I wanted to have clarity on. In principle the Scipy integrate function `dblquad` could be used on every pixel, but I have no idea if the performance would be acceptable. If this already had been shown to be 100 x slower, it would clearly not be an option, but otherwise probably needs further testing to see if it could be an alternative to oversampling.

As I said above, I did this test in IDL, not python, so I can not comment on the performance of the python integration routine. I doubt it will be comparably fast to a routine that simply plugs in the center of the pixel (or more precisely: its distance to a reference position) into a sersic profile, which is a significant shortcut.

> So you are saying the correct intensity at the exact location x, y (radius `r`, if this could indeed be measured somehow) is _not_ given by
> `Ie * np.exp(-bn * (r ** (1 / n) - 1))` ?

No, that's correct.
But it's only correct for that precise position.
Again, I do not know what the routine does exactly, but I would assume that (as everything else I have seen out there), the routine uses the CENTER of a pixel as the reference. This is not the correct value if the profile is CURVED.

> > Yes, it might not be a 'bug', because the code does what it's been told to do. It's just been told to do the wrong thing.
> 
> Well, it should return the intensity distribution as defined in the reference http://ned.ipac.caltech.edu/level5/March05/Graham/Graham2.html – if it doesn't even do that for sampled intensities, that is clearly a bug. 

I know the Graham paper. Btw, if this is the reference you use, it should also be made clear that there is another limitation:
bn = 1.9992n - 0.3271, is only true for 0.5 < n < 10 (no idea whether this is documented). There are other parameterisations0 that can go somewhat further.
If the routines assumes anything like a pixel center as the radius, it will most likely be wrong, yes.
Again, I don't know whether it does, but it is what the OP refers to and why he tagged me in this post.

>That it is not producing binned fluxes should at least be clarified in the documentation, but ideally I think models should provide either option, as it is e.g. implemented now for `Spectrum1D`. But that would require some major work on the entire `modeling` package.
> But from your comments above I take that this particular model should be taken care of by a galaxy person...

I would go further than that. It should not be used. 
Given that a sersic profile is only useful for galaxy persons, that would make the entire routine obsolete.

Technically, the workaround is simply: Produce the profile 1000 times larger, then rebin it such that the flux of 1000x1000 pixels go into 1 pixel. Done, with a oversampling of 1000, which according to my tests is sufficient WELL within the noise limit of any realistic images.
The reasons I have to go to different oversampling steps are two-fold:
- memory issues (large arrays need memory)
- speed issues (doing this over the entire profile out to large radii - several 10s or hundreds half-light radii - is slow and unnecessary.)

**Comment by dhomeier** [comment]
> I know the Graham paper. Btw, if this is the reference you use, it should also be made clear that there is another limitation:
> bn = 1.9992n - 0.3271, is only true for 0.5 < n < 10 (no idea whether this is documented). There are other parameterisations0 that can go somewhat further.

The actual implementation uses the inversion of Γ(2𝑛)=2𝛾(2𝑛,𝑏𝑛)
so should hopefully be safe.

> If the routines assumes anything like a pixel center as the radius, it will most likely be wrong, yes.
> Again, I don't know whether it does, but it is what the OP refers to and why he tagged me in this post.
> 
The routines do not assume anything about working on pixel centres or boundaries; they give function values for specific parameter values.

> > That it is not producing binned fluxes should at least be clarified in the documentation, but ideally I think models should provide either option, as it is e.g. implemented now for `Spectrum1D`. But that would require some major work on the entire `modeling` package.
> > But from your comments above I take that this particular model should be taken care of by a galaxy person...
> 
> I would go further than that. It should not be used.
> Given that a sersic profile is only useful for galaxy persons, that would make the entire routine obsolete.

I was referring to someone who could take care of working on this issue, which probably should be someone more familiar with how this specific model is typically used.
I don't think removing `Sersic2D` entirely would be seen as acceptable.

> Technically, the workaround is simply: Produce the profile 1000 times larger, then rebin it such that the flux of 1000x1000 pixels go into 1 pixel. Done, with a oversampling of 1000, which according to my tests is sufficient WELL within the noise limit of any realistic images.

For some very basic ballpark estimates on the Python implementations:
Evaluating the function on the inner 10 x 10 points at 100 times higher resolution (i.e. an array(1000, 1000)) took just as long as a _single_ call to `scipy.integrate.dblquad` (i.e. for 1 pixel) – that's not unexpected, as the integrator itself will make numerous calls to the function which are not themselves vectorised.
For 1000 times oversampling the time becomes slightly greater than integrating all 100 pixels with `dblquad`, but with an 800 MB input array I was clearly already running into cache limits here. So, a flexible oversampling scheme would certainly be more efficient.

**Comment by BorisHaeussler** [comment]
> I was referring to someone who could take care of working on this issue, which probably should be someone more familiar with how this specific model is typically used.
> I don't think removing `Sersic2D` entirely would be seen as acceptable.

No, probably not.
And yes, I understood that in the end, hence why I edited my comment (seemingly too late).
 
> For some very basic ballpark estimates on the Python implementations:
> Evaluating the function on the inner 10 x 10 points at 100 times higher resolution (i.e. an array(1000, 1000)) took just as long as a _single_ call to `scipy.integrate.dblquad` (i.e. for 1 pixel) – that's not unexpected, as the integrator itself will make numerous calls to the function which are not themselves vectorised.
> For 1000 times oversampling the time becomes slightly greater than integrating all 100 pixels with `dblquad`, but with an 800 MB input array I was clearly already running into cache limits here. So, a flexible oversampling scheme would certainly be more efficient.

Indeed.
Interesting to see the numbers.
It's indeed the reason why I went down that route myself years ago.

**Comment by dhomeier** [comment]
> > Evaluating the function on the inner 10 x 10 points at 100 times higher resolution (i.e. an array(1000, 1000)) took just as long as a _single_ call to `scipy.integrate.dblquad` (i.e. for 1 pixel) – that's not unexpected, as the integrator itself will make numerous calls to the function which are not themselves vectorised.
> > For 1000 times oversampling the time becomes slightly greater than integrating all 100 pixels with `dblquad`, but with an 800 MB input array I was clearly already running into cache limits here. So, a flexible oversampling scheme would certainly be more efficient.
> 
> Indeed.
> Interesting to see the numbers.
> It's indeed the reason why I went down that route myself years ago.

A follow-up on the timing tests:
Running all tests on the central 100 (`img[45:55, 45:55]`) points of the example model from https://docs.astropy.org/en/latest/api/astropy.modeling.functional_models.Sersic2D.html, average times and maximum relative/absolute deviations from integrated solution on a 16 GB i7 MB Pro were:
- 100 times oversampling: 65 ms, 0.0175, 1.0
- 1000 times oversampling: 11 s, 0.0015, 0.09
 - dblquad: 22 s, 0, 0
 - dblquad(epsrel=1e-4, epsabs=1e-3): 8.6 s, 2.4e-8, 2.1e-6
 
So especially when relaxing the accuracy setting a bit, the optimised Scipy `dblquad` is very competitive with high oversampling – at still much greater accuracy and without the memory impact. But in practice, a combination with moderate oversampling for the intermediate region will still be required.

**Comment by canorve** [comment]
> I'd be happy to provide my own code if you want to have a look, but it's written in IDL, so not very useful directly.

Hi Boris, can I see your code? I think the code was on the https://www.mpia.de/GEMS/gems.htm web page years ago, but that link does not work anymore.

**Comment by hamogu** [comment]
A more general comment: In [Sherpa](https://github.com/sherpa/sherpa/) data objects and models have an "integrate" flag. If False, they work like astropy models (model is evaluated at (x,y), if True, the data has to be specified with bin edges (x_low, x_hi, y) and the model is evaluated as "integral from x_lo to x_hi". For models that have an analytical solution for the integral (e.g. polynomials, Gaussians), the model is essentially implemented twice and the integrate flag decides if the formula for the model or for the integral of the model is used. Models that don't have an analytical integral could be integrated numerically.

That's a route that astropy could go, too, but as @dhomeier said above it's a pretty major undertaking, particularly if you want to make sure it works in n dimensions, not just in 1 or 2.

**Comment by dhomeier** [comment]
I didn't think there were >2D models at this point; still an efficient implementation would probably need to be tuned to the characteristics of the various model. E.g. numerical integration with `scipy.integrate` is hardly a viable option for a (100, 100) array, much less for a 1k image, so integrating models like `Sersic2D` will need a more flexible approach.

**Comment by Knusper** [comment]
Sorry for adding more noise here:  This affects all models in Astropy. When pixel binning needs to be taken into account, e.g. on CCD images, then the way to go might be to convolve the profiles with a box-car of pixel size... shouldn't that do the trick?  These numerical "we rebin finer in the centre" schemes used in Galfit and elsewhere seem always a bit blunt...

**Comment by Knusper** [comment]
In my recent paper - https://doi.org/10.1002/asna.20220091 - I also caution against using the Sersic2D model in astropy to actually model observed data because of this issue (footnote 10 on p. 24).   This really should be addressed, since the idea of the "modelling" class is to "model" something.  Again, its not only the Sersic model, but all models do only sample at the pixel centre.  However, in astropy.convolution the user has the choice to properly deal with sampling issues.
